### PR TITLE
Add support for collection expressions in attributes

### PR DESCRIPTION
**Pull request by jcouv**
Fixes https://github.com/dotnet/roslyn/issues/69133

### CODE REVIEW

File: src/Compilers/CSharp/Test/Emit2/Semantics/CollectionExpressionTests.cs [hunk 1]
```diff
@@ -14755,5 +14755,277 @@ static void Main(List<int> list)
...
+        [Fact, WorkItem("https://github.com/dotnet/roslyn/issues/69133")]
+        public void InAttribute_Empty()
+        {
+            string source = """
+                using System;
+                
+                var attr = (XAttribute)System.Attribute.GetCustomAttribute(typeof(C), typeof(XAttribute));
+                Console.Write(attr._values.Length);
+                
+                [X([])]
```
Comment (CyrusNajmabadi) [inline src/Compilers/CSharp/Test/Emit2/Semantics/CollectionExpressionTests.cs:14801]: test with a string in the literal, and 'null' in the literal?

File: src/Compilers/CSharp/Test/Emit2/Semantics/CollectionExpressionTests.cs [hunk 2]
```diff
@@ -14755,5 +14755,277 @@ static void Main(List<int> list)
...
+
+                public class XAttribute : System.Attribute
+                {
+                    public XAttribute(int[] values) { }
+                }
+                """;
+
+            CreateCompilation(source).VerifyEmitDiagnostics(
+                // (1,14): error CS9176: There is no target type for the collection expression.
+                // [X([1, 2, .. [3]])]
```
Comment (CyrusNajmabadi) [inline src/Compilers/CSharp/Test/Emit2/Semantics/CollectionExpressionTests.cs:14857]: definitely an odd error :)

File: src/Compilers/CSharp/Test/Emit2/Semantics/CollectionExpressionTests.cs [hunk 3]
```diff
@@ -14755,5 +14755,277 @@ static void Main(List<int> list)
...
+                
+                [X([])]
+                class C
+                {
+                }
+                
+                public class XAttribute : System.Attribute
+                {
+                    public int[] _values;
+                    public XAttribute(int[] values) { _values = values; }
```
Comment (CyrusNajmabadi) [inline src/Compilers/CSharp/Test/Emit2/Semantics/CollectionExpressionTests.cs:14809]: test with `param int[]`?

File: src/Compilers/CSharp/Test/Emit2/Semantics/CollectionExpressionTests.cs (hunk 2 above)
Comment (jcouv) [reply src/Compilers/CSharp/Test/Emit2/Semantics/CollectionExpressionTests.cs:14857]: I agree. It's an existing problem though and I'm not planning to address in this PR.

File: src/Compilers/CSharp/Test/Emit2/Semantics/CollectionExpressionTests.cs [hunk 4]
```diff
@@ -17175,5 +17175,349 @@ static void Main(List<int> list)
...
+                }
+                """;
+
+            var comp = CreateCompilation(sourceA, targetFramework: TargetFramework.Net80);
+            comp.VerifyDiagnostics(
+                // (12,2): error CS0181: Attribute constructor parameter 'a' has type 'A', which is not a valid attribute parameter type
+                // [X([1])]
+                Diagnostic(ErrorCode.ERR_BadAttributeParamType, "X").WithArguments("a", "A").WithLocation(12, 2)
+                );
+        }
```
Comment (cston) [inline src/Compilers/CSharp/Test/Emit2/Semantics/CollectionExpressionTests.cs:17521]: >} [](http://example.com/codeflow?start=8&length=1)

Please test a spread: `[X([.. new List<int>()])]`

File: src/Compilers/CSharp/Test/Emit2/Semantics/CollectionExpressionTests.cs [hunk 5]
```diff
@@ -17175,5 +17175,349 @@ static void Main(List<int> list)
...
+                    public XAttribute(int[][] values) { }
+                }
+                """;
+
+            var comp = CreateCompilation(source, targetFramework: TargetFramework.Net70).VerifyEmitDiagnostics(
+                // (1,2): error CS0181: Attribute constructor parameter 'values' has type 'int[][]', which is not a valid attribute parameter type
+                // [X([[1], [2]])]
+                Diagnostic(ErrorCode.ERR_BadAttributeParamType, "X").WithArguments("values", "int[][]").WithLocation(1, 2)
+                );
+        }
```
Comment (cston) [inline src/Compilers/CSharp/Test/Emit2/Semantics/CollectionExpressionTests.cs:17282]: >} [](http://example.com/codeflow?start=8&length=1)

Consider testing `[X([(int[])[1]])]` with `XAttribute(object[] values) { }`

File: src/Compilers/CSharp/Test/Emit2/Semantics/CollectionExpressionTests.cs (hunk 4 above)
Comment (jcouv) [reply src/Compilers/CSharp/Test/Emit2/Semantics/CollectionExpressionTests.cs:17521]: Good catch. Thanks

**Comment by jaredpar** [comment]
@jcouv make sure to retarget this to `release/dev17.8`

File: src/Compilers/CSharp/Test/Emit2/Semantics/CollectionExpressionTests.cs [hunk 6]
```diff
@@ -17175,5 +17175,421 @@ static void Main(List<int> list)
...
+                {
+                }
+
+                public class XAttribute : System.Attribute
+                {
+                    public XAttribute(int[][] values) { }
+                }
+                """;
+
+            CreateCompilation(source, targetFramework: TargetFramework.Net70).VerifyEmitDiagnostics(
```
Comment (cston) [inline src/Compilers/CSharp/Test/Emit2/Semantics/CollectionExpressionTests.cs:17277]: >targetFramework: TargetFramework.Net70 [](http://example.com/codeflow?start=38&length=38)

Is there a reason to use `TargetFramework.Net70` for this test and several below?

File: src/Compilers/CSharp/Test/Emit2/Semantics/CollectionExpressionTests.cs [hunk 7]
```diff
@@ -17175,5 +17175,421 @@ static void Main(List<int> list)
...
+                );
+        }
+
+        [Fact, WorkItem("https://github.com/dotnet/roslyn/issues/69133")]
+        public void InAttribute_NestedArrayAsObject()
+        {
+            string source = """
+                var attr = (XAttribute)System.Attribute.GetCustomAttribute(typeof(C), typeof(XAttribute));
+                var inner = (int[])attr._values[0];
+                System.Console.Write($"OuterLength={attr._values.Length} InnerValue={inner[0]} InnerLength={inner.Length}");
```
Comment (cston) [inline src/Compilers/CSharp/Test/Emit2/Semantics/CollectionExpressionTests.cs:17290]: Consider using `attr._values.Report();` in this test and below.

**Comment by CyrusNajmabadi** [comment]
I haven't checked, but are there tests for using a collection-expr as an argument to an attribute 'named arg' and an attribute 'named field/prop'?

**Comment by jcouv** [comment]
> I haven't checked, but are there tests for using a collection-expr as an argument to an attribute 'named arg' and an attribute 'named field/prop'?

Thanks, that's a great suggestion. Not sure how I missed that. Will add some tests

File: src/Compilers/CSharp/Portable/Binder/Binder_Attributes.cs [hunk 8]
```diff
@@ -989,6 +999,13 @@ private TypedConstant VisitConversion(BoundConversion node, BindingDiagnosticBag
                 var operand = node.Operand;
                 var operandType = operand.Type;
 
+                if (node.Conversion.IsCollectionExpression
+                    && node.Conversion.GetCollectionExpressionTypeKind(out _) == CollectionExpressionTypeKind.Array)
```
Comment (333fred) [inline src/Compilers/CSharp/Portable/Binder/Binder_Attributes.cs:1003]: This seems likely to cause a binding cycle. `GetCollectionExpressionTypeKind` calls `GetEarlyDecodedWellKnownAttributeData`, which forces binding of early attributes, which we're already in the middle of doing here. `GetCollectionExpressionTypeKind` should likely be updated to have an `isEarlyBinding` flag and avoid doing anything but checking for array in such cases.

File: src/Compilers/CSharp/Test/Emit2/Semantics/CollectionExpressionTests.cs
Comment (333fred) [inline src/Compilers/CSharp/Test/Emit2/Semantics/CollectionExpressionTests.cs:1]: Please add some recursive tests where attributes are applied to themselves. Make sure to include some tests with `CollectionBuilder` being applied to itself and being passed a collection expression.

**Review by 333fred — COMMENTED** [review]
Done review pass

File: src/Compilers/CSharp/Portable/Binder/Binder_Attributes.cs (hunk 8 above)
Comment (jcouv) [reply src/Compilers/CSharp/Portable/Binder/Binder_Attributes.cs:1003]: This is `Conversion.GetCollectionExpressionTypeKind` which doesn't pull on `GetEarlyDecodedWellKnownAttributeData`.

**Review by 333fred — APPROVED** [review]
Thanks for the additional tests. I do think we likely have a cycle here, but it's not introduced by this PR, probably pre-existing.

**Comment by CyrusNajmabadi** [comment]
Woohoo!  Thanks for getting this in!

**Pull request merged**